waku/v2/protocol/waku_noise/noise.nim

```diff
@@ -917,4 +923,565 @@ proc deserializePayloadV2*(payload: seq[byte]): Result[PayloadV2, cstring]
+          outHandshakeMessage.add NoisePublicKey(flag: 1, pk: encS)
+        else:
+          outHandshakeMessage.add NoisePublicKey(flag: 0, pk: encS)
+
+    of T_psk:
+
+      # If user is reading the "psk" token
+
+      trace "noise psk"
+
+      # Calls MixKeyAndHash(psk)
+      hs.ss.mixKeyAndHash(hs.psk)
+
+    of T_ee:
+
+      # If user is reading the "ee" token
+
+      trace "noise dh ee"
+
+      # If local and/or remote ephemeral keys are not set, we raise an error
+      if hs.e == default(KeyPair) or hs.re == default(Curve25519Key):
+        raise newException(NoisePublicKeyError, "Local or remote ephemeral key not set")
+
+      # Calls MixKey(DH(e, re)).
+      hs.ss.mixKey(dh(hs.e.privateKey, hs.re))
+
+    of T_es:
+
+      # If user is reading the "es" token
+
+      trace "noise dh es"
+
+      # We check if keys are correctly set.
+      # If both present, we call MixKey(DH(e, rs)) if initiator, MixKey(DH(s, re)) if responder.
+      if hs.initiator:
+        if hs.e == default(KeyPair) or hs.rs == default(Curve25519Key):
+          raise newException(NoisePublicKeyError, "Local or remote ephemeral/static key not set")
+        hs.ss.mixKey(dh(hs.e.privateKey, hs.rs))
+      else:
+        if hs.re == default(Curve25519Key) or hs.s == default(KeyPair):
+          raise newException(NoisePublicKeyError, "Local or remote ephemeral/static key not set")
+        hs.ss.mixKey(dh(hs.s.privateKey, hs.re))
+
+    of T_se:
+
+      # If user is reading the "se" token
+
+      trace "noise dh se"
+
+      # We check if keys are correctly set.
+      # If both present, call MixKey(DH(s, re)) if initiator, MixKey(DH(e, rs)) if responder.
+      if hs.initiator:
+        if hs.s == default(KeyPair) or hs.re == default(Curve25519Key):
+          raise newException(NoiseMalformedHandshake, "Local or remote ephemeral/static key not set")
+        hs.ss.mixKey(dh(hs.s.privateKey, hs.re))
+      else:
+        if hs.rs == default(Curve25519Key) or hs.e == default(KeyPair):
+          raise newException(NoiseMalformedHandshake, "Local or remote ephemeral/static key not set")
+        hs.ss.mixKey(dh(hs.e.privateKey, hs.rs))
+
+    of T_ss:
+
+      # If user is reading the "ss" token
+
+      trace "noise dh ss"
+
+      # If local and/or remote static keys are not set, we raise an error
+      if hs.s == default(KeyPair) or hs.rs == default(Curve25519Key):
+        raise newException(NoiseMalformedHandshake, "Local or remote static key not set")
+
+      # Calls MixKey(DH(s, rs)).
+      hs.ss.mixKey(dh(hs.s.privateKey, hs.rs))
+
+  return ok(outHandshakeMessage)
+
+#################################
+## Procedures to progress handshakes between users
+#################################
+
+# Initializes a Handshake State
+proc initialize*(hsPattern: HandshakePattern, staticKey: KeyPair = default(KeyPair), prologue: seq[byte] = @[], psk: seq[byte] = @[], preMessagePKs: seq[NoisePublicKey] = @[], initiator: bool = false): HandshakeState 
```

Comment:
As per our 1:1, you may add an ephemeral keypair as a parameter here to account for pre-shared ephemeral keys